apps/omg_child_chain/lib/omg_child_chain/gas_price/history/server.ex

```diff
@@ -0,0 +1,160 @@
+    event_bus = Keyword.fetch!(opts, :event_bus)
+    :ok = event_bus.subscribe({:root_chain, "ethereum_new_height"}, link: true)
+
+    # The ets table is not initialized with `:read_concurrency` because we are expecting interleaving
+    # reads and writes. See http://erlang.org/doc/man/ets.html
+    history_ets = :ets.new(@history_table, [:ordered_set, :protected, :named_table])
+
+    state = %__MODULE__{
+      num_blocks: num_blocks,
+      history_ets: history_ets
+    }
+
+    _ = Logger.info("Started #{__MODULE__}: #{inspect(state)}")
+    {:ok, state, {:continue, :first_fetch}}
+  end
+
+  @doc false
+  @impl GenServer
+  def handle_continue(:first_fetch, state) do
+    {:ok, to_height} = EthereumHeight.get()
+    from_height = to_height - state.num_blocks
+    {:ok, earliest_height, latest_height} = do_populate_prices(from_height, to_height, state)
+
+    state = %{state | earliest_stored_height: earliest_height, latest_stored_height: latest_height}
+
+    {:noreply, state}
+  end
+
+  #
+  # GenServer callbacks
+  #
+
+  @doc false
+  @impl GenServer
+  def handle_info({:internal_event_bus, :ethereum_new_height, height}, state) do
+    from_height = height - state.num_blocks
+    {:ok, earliest_height, latest_height} = do_populate_prices(from_height, height, state)
+
+    {:noreply, %{state | earliest_stored_height: earliest_height, latest_stored_height: latest_height}}
+  end
+
+  @doc false
+  @impl GenServer
+  def handle_cast({:subscribe, subscriber}, state) do
+    subscribers = Enum.uniq([subscriber | state.subscribers])
+    {:noreply, %{state | subscribers: subscribers}}
+  end
+
+  #
+  # Internal implementations
+  #
+
+  defp do_populate_prices(from_height, to_height, state) do
+    fetch_from_height = max(from_height, state.latest_stored_height)
+
+    # Fetch and insert new heights, leaving obsolete heights intact.
+    :ok =
+      fetch_from_height..to_height
+      |> Fetcher.stream()
+      |> stream_insert(state.history_ets)
+      |> Stream.run()
+
+    # Prune obsolete heights.
+    :ok = prune_heights(state.history_ets, state.earliest_stored_height, from_height - 1)
+
+    _ =
+      Logger.info(
+        "#{__MODULE__} removed gas prices from Eth heights: #{state.earliest_stored_height} - #{from_height - 1}."
+      )
+
+    _ = Logger.info("#{__MODULE__} available gas prices from Eth heights: #{from_height} - #{to_height}.")
+
+    # Inform all subscribers that the history has been updated.
+    _ = Enum.each(state.subscribers, fn subscriber -> send(subscriber, {History, :updated}) end)
+    {:ok, from_height, to_height}
+  end
+
+  defp stream_insert(stream_blocks, history_ets) do
+    Stream.each(stream_blocks, fn block ->
+      height = Encoding.int_from_hex(block["number"])
+      prices = Enum.map(block["transactions"], fn tx -> Encoding.int_from_hex(tx["gasPrice"]) end)
+      timestamp = Encoding.int_from_hex(block["timestamp"])
+
+      true = :ets.insert(history_ets, {height, prices, timestamp})
```

Comment:
It's already streaming per item though (because of the processing needed on each item). Not sure if it's useful to bulk it the stream up before inserting... 🤔